processing/src/main/java/org/apache/druid/query/monomorphicprocessing/SpecializationService.java

```diff
@@ -65,19 +69,155 @@
+    //  }
+
+    // this is getting meta
+    MethodHandle defineClass = lookup.unreflect(MethodHandles.Lookup.class.getMethod("defineClass", byte[].class));
+    MethodHandle privateLookupIn = lookup.findStatic(
+        MethodHandles.class,
+        "privateLookupIn",
+        MethodType.methodType(MethodHandles.Lookup.class, Class.class, MethodHandles.Lookup.class)
+    );
+
+    // bind privateLookupIn lookup argument to this method's lookup
+    privateLookupIn = MethodHandles.insertArguments(privateLookupIn, 1, lookup);
+
+    // -> defineClass(Class targetClass, byte[] byteCode)
+    defineClass = MethodHandles.filterArguments(defineClass, 0, privateLookupIn);
+
+    // add a dummy String argument to match the corresponding JDK8 version
+    // -> defineClass(Class targetClass, byte[] byteCode, String className)
+    defineClass = MethodHandles.dropArguments(defineClass, 2, String.class);
+    return defineClass;
+  }
+
+  private static MethodHandle defineClassJava8(MethodHandles.Lookup lookup) throws ReflectiveOperationException
+  {
+    Class<?> unsafeClass = Class.forName("sun.misc.Unsafe");
+    Field f = unsafeClass.getDeclaredField("theUnsafe");
+    f.setAccessible(true);
+    Object theUnsafe = f.get(null);
+
+    // "Compile" a MethodHandle that is equilavent to
+    //
+    //  Class<?> defineClass(Class targetClass, byte[] byteCode, String className) {
+    //    return Unsafe.defineClass(
+    //        className,
+    //        byteCode,
+    //        0,
+    //        byteCode.length,
+    //        targetClass.getClassLoader(),
+    //        targetClass.getProtectionDomain()
+    //    );
+    //  }
+
+    MethodHandle defineClass = lookup.findVirtual(
+        unsafeClass,
+        "defineClass",
+        MethodType.methodType(
+            Class.class,
+            String.class,
+            byte[].class,
+            int.class,
+            int.class,
+            ClassLoader.class,
+            ProtectionDomain.class
+        )
+    ).bindTo(theUnsafe);
+
+    MethodHandle getProtectionDomain = lookup.unreflect(Class.class.getMethod("getProtectionDomain"));
+    MethodHandle getClassLoader = lookup.unreflect(Class.class.getMethod("getClassLoader"));
+
+    // apply methods to the targetClass
+    // -> defineClass(String className, byte[] byteCode, int offset, int length, Class targetClass, Class targetClass)
+    defineClass = MethodHandles.filterArguments(defineClass, 5, getProtectionDomain);
+    defineClass = MethodHandles.filterArguments(defineClass, 4, getClassLoader);
+
+    // duplicate the last argument to apply the methods above to the same class
+    // -> defineClass(String className, byte[] byteCode, int offset, int length, Class targetClass, Class targetClass)
+    defineClass = MethodHandles.permuteArguments(
+        defineClass,
+        MethodType.methodType(Class.class, String.class, byte[].class, int.class, int.class, Class.class),
+        0, 1, 2, 3, 4, 4
```

Comment:
Could you add comments to these constants? I don't understand how do see work, even knowing that they are related to "permutation" somehow.

Reply:
I've updated the comments to explain in short lambda expressions how it transforms the MethodHandle, and what the code does to the method call, hopefully that makes it clearer. The index values refer to argument positions in the resulting method handle, and their ordering defines the order in which they are applied to the target method handle. Duplicate values have the effect of passing the same argument twice to the target handle, as is the case here with the targetClass argument.

Reply:
Each argument should be on its own line

Reply:
I think this format makes it readable. We typically don't break up array declarations or map key/value pairs either for that reason.

Reply:
I agree that it's more readable to keep them in one line in this case, but when we will figure out how to enforce the rule with Checkstyle, this will be reformatted anyway. (We have to be able to set up exclusions for methods like `ImmutableMap.of()`, but probably not such rarely used ones as `MethodHandles.permuteArguments()`.) Consider this comment optional.